dev/update_mlflow_versions.py

```diff
@@ -26,101 +61,101 @@ def replace_occurrences(files: List[Path], pattern: Union[str, re.Pattern], repl
 
-    # Java
-    for java_extension in ["java"]:
-        replace_occurrences(
-            files=Path("mlflow", "java").rglob(f"*.{java_extension}"),
-            pattern=rf"{re.escape(current_py_version_without_suffix)}(-SNAPSHOT)?",
-            repl=replace_dev_suffix_with(new_py_version, "-SNAPSHOT"),
-        )
 
-    for xml_extension in ["xml"]:
-        # the pom.XML files define versions of dependencies as well.
-        # this causes issues when the mlflow version matches the
-        # version of a dependency. to work around, we make sure to
-        # match only the correct keys
-        old_py_version_pattern = rf"{re.escape(current_py_version_without_suffix)}(-SNAPSHOT)?"
-        dev_suffix_replaced = replace_dev_suffix_with(new_py_version, "-SNAPSHOT")
-
-        # group 1: everything before the version
-        # group 2: optional -SNAPSHOT
-        # group 3: everything after the version
-        replace_str = f"\\g<1>{dev_suffix_replaced}\\g<3>"
-
-        mlflow_version_tag_pattern = rf"""(<mlflow.version>){
-            old_py_version_pattern
-        }(</mlflow.version>)"""
-        replace_occurrences(
-            files=Path("mlflow", "java").rglob(f"*.{xml_extension}"),
-            pattern=mlflow_version_tag_pattern,
-            repl=replace_str,
-        )
+def replace_java(old_version: str, new_py_version: str, paths: List[Path]) -> None:
+    old_py_version_pattern = get_java_py_version_pattern(old_version)
+    dev_suffix_replaced = get_java_new_py_version(new_py_version)
 
-        mlflow_parent_pattern = rf"""(<artifactId>mlflow-parent</artifactId>\s+<version>){
-            old_py_version_pattern
-        }(</version>)"""
-        replace_occurrences(
-            files=Path("mlflow", "java").rglob(f"*.{xml_extension}"),
-            pattern=mlflow_parent_pattern,
-            repl=replace_str,
-        )
+    replace_occurrences(
+        files=paths,
+        pattern=old_py_version_pattern,
+        repl=dev_suffix_replaced,
+    )
+
+
+# Note: the pom.xml files define versions of dependencies as
+# well. this causes issues when the mlflow version matches the
+# version of a dependency. to work around, we make sure to
+# match only the correct keys
+def replace_java_pom_xml(old_version: str, new_py_version: str, paths: List[Path]) -> None:
+    old_py_version_pattern = get_java_py_version_pattern(old_version)
+    dev_suffix_replaced = get_java_new_py_version(new_py_version)
+
+    mlflow_version_tag_pattern = r"<mlflow.version>"
+    mlflow_spark_pattern = (
+        r"<artifactId>mlflow-spark_\${scala\.compat\.version}</artifactId>\s+<version>"
+    )
+    mlflow_parent_pattern = r"<artifactId>mlflow-parent</artifactId>\s+<version>"
+
+    # combine the three tags together to form the regex
+    mlflow_replace_pattern = (
+        rf"({mlflow_version_tag_pattern}|{mlflow_spark_pattern}|{mlflow_parent_pattern})"
+        + f"{old_py_version_pattern}"
+        + r"(</mlflow.version>|</version>)"
+    )
```

Comment:
i also refactored the XML code, instead of running it 3 times we can just combine three regexes into one